### PR TITLE
Add support for dumping and using precise debug info

### PR DESCRIPTION
**Pull request by jakobbotsch**
* In the JIT, add support for dumping the precise debug info out through
  an environment variable `DOTNET_JitDumpPreciseDebugInfoFile` in a
  simple JSON format. This is a stopgap until we expose the extra
  information through ETW events.

* In dotnet-pgo, add an argument --precise-debug-info-file which can
  point to the file produced by the JIT. When used, dotnet-pgo will get
  native<->IL mappings from this file instead of through ETW events.

* In dotnet-pgo, add support for attributing samples to inlinees when
  that information is present. This changes the attribution process a
  bit: previously, we would group all LBR data/samples and then
  construct the profile from all the data. We now do it in a more
  streaming way where there is a SampleCorrelator that can handle
  individual LBR records and individual samples.

* In dotnet-pgo, add an argument --dump-worst-overlap-graphs-to which
  can be used in the compare-mibc command to dump out a .dot file
  containing the flow graph of the methods with the worst overlap
  measures, and showing the relative weight count on each basic block
  and edge for the two profiles being compared. This is particular
  useful to find out where we are producing incorrect debug mappings, by
  comparing spgo.mibc and instrumented.mibc files.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
@AndyAyersMS @davidwrighton Can you please take a look? No need to review the SPGO parts in detail yet since it's more or less at a prototype stage.

File: src/coreclr/jit/codegencommon.cpp [hunk 1]
```diff
@@ -10801,6 +10806,102 @@ void CodeGen::genIPmappingGen()
...
+
+void CodeGen::genDumpPreciseDebugInfo()
+{
+    if (JitConfig.JitDumpPreciseDebugInfoFile() == nullptr)
+        return;
+
+    static unsigned int s_flag;
+
+    while (InterlockedCompareExchange(&s_flag, 1, 0) != 0)
+        System_YieldProcessor();
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegencommon.cpp:10849]: Do we have a better lock primitive in the JIT I can use here?

**Comment by jakobbotsch** [comment]
The dumped flow graphs look like this:
image

This is very helpful to see that clearly the attribution process is not working well for this particular function.

File: src/coreclr/jit/codegencommon.cpp (hunk 1 above)
Comment (SingleAccretion) [reply src/coreclr/jit/codegencommon.cpp:10849]: There is [`CritSecHolder`](https://github.com/dotnet/runtime/blob/7afb66262c287b7832b872473d1e668038aa2496/src/coreclr/jit/utils.h#L741), if that would be suitable here.

Comment (jakobbotsch) [reply src/coreclr/jit/codegencommon.cpp]: That looks better, thanks.

File: src/coreclr/tools/dotnet-pgo/PgoCompareMethodFlowGraph.cs [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        public bool ProfilesHadEdges { get; init; }
+        public PgoCompareMethodBasicBlock EntryBasicBlock { get; init; }
+        public List<PgoCompareMethodBasicBlock> BasicBlocks { get; init; }
+
+        public long TotalBlockCount1 => BasicBlocks.Sum(bb => bb.BlockCount1);
+        public long TotalBlockCount2 => BasicBlocks.Sum(bb => bb.BlockCount2);
+        public long TotalEdgeCount1 => BasicBlocks.Sum(bb => bb.Edges.Sum(e => e.Value.Count1));
+        public long TotalEdgeCount2 => BasicBlocks.Sum(bb => bb.Edges.Sum(e => e.Value.Count2));
+
+        public double ComputeBlockOverlap()
```
Comment (AndyAyersMS) [inline src/coreclr/tools/dotnet-pgo/PgoCompareMethodFlowGraph.cs:26]: You might also look into the Spearman rank correlation as a measure of similarity.

File: src/coreclr/jit/codegencommon.cpp [hunk 3]
```diff
@@ -10801,6 +10806,98 @@ void CodeGen::genIPmappingGen()
...
+        fprintf(file, "]}");
+    }
+}
+
+void CodeGen::genDumpPreciseDebugInfo()
+{
+    if (JitConfig.JitDumpPreciseDebugInfoFile() == nullptr)
+        return;
+
+    static CritSecObject s_critSect;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/codegencommon.cpp:10846]: Generally we don't want the jit to take locks but using one during bring-up like this is ok.

File: src/coreclr/tools/dotnet-pgo/SPGO/SampleProfile.cs [hunk 4]
```diff
@@ -16,69 +16,56 @@ namespace Microsoft.Diagnostics.Tools.Pgo
...
+            Debug.Assert(FlowGraph.Lookup(bb.Start) == bb);
+            CollectionsMarshal.GetValueRefOrAddDefault(_rawSamples, bb, out _) += count;
+            AttributedSamples += count;
         }
 
-        /// <summary>
-        /// Given some IL offset samples into a method, construct a profile.
-        /// </summary>
-        public static SampleProfile Create(MethodIL il, FlowGraph fg, IEnumerable<int> ilOffsetSamples)
+        public void SmootheFlow()
```
Comment (AndyAyersMS) [inline src/coreclr/tools/dotnet-pgo/SPGO/SampleProfile.cs:55]: Nit ..?
```suggestion
        public void SmoothFlow()
```

**Review by AndyAyersMS — APPROVED** [review]
LGTM.

Glad to see this moving along.

File: src/coreclr/tools/dotnet-pgo/SPGO/SampleProfile.cs (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/SPGO/SampleProfile.cs]: Today I learned that 'smooth' can also be used as a verb. :-)

File: src/coreclr/tools/dotnet-pgo/PgoCompareMethodFlowGraph.cs (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/tools/dotnet-pgo/PgoCompareMethodFlowGraph.cs:26]: I'll look into it.

**Comment by jakobbotsch** [comment]
Gonna merge this as I have some follow-up work that builds on it. @davidwrighton let me know if you have any feedback and I'll address it separately.

**Pull request merged**